independent-projects/bootstrap/runner/src/main/java/io/quarkus/bootstrap/runner/JarFileReference.java

```diff
@@ -205,6 +194,12 @@ private static CompletableFuture<JarFileReference> syncLoadAcquiredJarFile(JarRe
         }
     }
 
+    private static CompletableFuture<JarFileReference> completedWith(JarFile jarFile) {
```

Comment:
This could be changed to use `completeWith` - in order to not make it complex in the mind of the readers that it has slightly different behaviours in term of the Java Memory Model

i.e. 
1. we create the `completedFuture` but it's instance is not yet exposed (i.e. setting an atomic reference to hold it)
2. `jarFileRef.completedFuture` is not using a volatile and is just a plain field, but here is not `stored released` by the completion
3. it relies on the caller which is using  a `AtomicReference::compareAndSet`  to `store release it`

It means than that who read this field, here, need to `load acquire` the fied through `AtomicReference::get`, while. on `completeWith` it just need to perform `CompletableFuture::getNow` or `CompletableFuture::join` instead.
Thanksfully both operations are required to obtain the `jarFileRef` instance and close/release it, making both correct `synchronized with`/`happens before` relations.

Using a single mechanism as suggested will make easier to think about this and unify the approach, despite the original code should still be correct, concurrency-wise

Reply:
I removed both those methods and moved the logic in the constructor. This also allowed to declare the `completedFuture` field as final. @franz1981 please give a second look.